Docker Playground Feedback

knaufk opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

Myasuka committed:
[FLINK-11860][config] Update usages of deprecated (job-/taskmanager).heap.mb

Commit:
Revert "[FLINK-8801][yarn/s3] fix Utils#setupLocalResource() relying on consistent read-after-write"

This reverts commit c90a757b29f168144b1bae99df532911ae682e63.

Commit:
[FLINK-8801][yarn/s3] Retry if remote file not found and use remote modification time

zentol committed:
[hotfix][build] Use HTTPS URL in qa-check.sh

zentol committed:
[hotfix][py] Use HTTPS URLs in package information

zentol committed:
[hotfix][build] Use HTTPS URLs

yngwiewang committed:
[FLINK-13227][docs-zh] Translate "asyncio" page into Chinese

This closes #9150

Myasuka committed:
[FLINK-12916][tests] Retry cancelWithSavepoint on cancellation barrier

knaufk committed:
[FLINK-12749] [docs] add "Docker Playgrounds" subsection to "Getting Started" documention

sjwiesman committed:
[FLINK-12747][docs] Getting Started - Table Api Walkthrough

danny0405 committed:
[FLINK-13074][table-planner] Fix PartitionableTableSink doesn't work for flink and blink planner

This closes #8966

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -0,0 +1,27 @@
+---
+title: "Flink Cluster Playground"
+nav-title: 'Flink Cluster Playground'
+nav-parent_id: docker-playgrounds
```
https://fontawesome.com/icons/docker?style=brands ?

knaufk replied:
I did not want to get into this: "This work includes material that may be protected as a trademark in some jurisdictions. If you want to use it, you have to ensure that you have the legal right to do so and that you do not infringe any trademark rights.
The use of these trademarks does not indicate endorsement of the trademark holder by Font Awesome, nor vice versa.
See the trademark owner for rules about appropriate use of their trademarks."

See also: https://github.com/wesbos/Font-Awesome-Docker-Icon. I am not comfortable in using it.

sjwiesman commented on docs/getting-started/docker-playgrounds/docker-compose.yaml:
```diff
@@ -0,0 +1,50 @@
+version: "2.1"
```
So this file is temporary or we will include a download link?

knaufk replied:
Yes, I thought we'll only embedded into the documentation. Download would of course also work. What do you think?

sjwiesman replied:
Ideally, I would like to, but if we can't get it into the CI pipeline I'm sure it will quickly and quietly get stale and go out of sync.

knaufk replied:
Not sure I completely understand your last comment.

In the long run we will only need the docker-compose file (when Flink Configuration can be injected into the containers by environment variables). I don't think there is any way around manually maintaining this file. Once this is only one file, it is also easier to simply put it somewhere to download. Currently, we would need an archive.

One advantage of embedding it into the docs is, that it we can template things like the Flink version.

knaufk replied:
Alternatively, I can of course put these files into a separate repository (e.g. flink-playgrounds/flink-cluster-playground). This would be the easiest for the user, but is of course also not optimal.

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
````diff
@@ -25,3 +25,183 @@ under the License.
+docker-compose logs -f taskmanager
+```
+
+# Operational Plays
````
Some other interesting operational tasks. Will add as I think of things. 

1) curl metrics - Keeps it light weight but we can still discuss metrics systems

knaufk replied:
Good point. I also want to make this REST & CLI whever possible (similar to the Scala/Java-tabs).

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
````diff
@@ -25,3 +25,183 @@ under the License.
+
+Rescale from Savepoint
+```
+docker-compose run --no-deps client flink run -p 3 -s <savepoint-path> -d /opt/flink/examples/streaming/StateMachineExample.jar --error-rate 0.01 --sleep 10 --backend file --checkpoint-dir file:///tmp/flink-checkpoint-directory
````
`--error-rate 0.01` I'm guessing this means the job will periodically fail and restart from checkpoint?

knaufk replied:
No, it will only print and "Alert" to STDOUT. No failures.

sjwiesman commented on docs/getting-started/docker-playgrounds/flink_cluster_playground.md:
```diff
@@ -25,3 +25,183 @@ under the License.
+  flink-checkpoint-directory:
+{% endhighlight %}
+
+**log4j-cli.properties**
```
When writing we'll probably want a sentence describing why two logger properties. It might not be obvious to new users from the file names.

knaufk replied:
Yes, btw: these are the default log4j properties of Flink, but since we mount the conf directory we override the ones that came with the image.

sjwiesman commented on docs/getting-started/docker-playgrounds/conf/flink-conf.yaml:
```diff
@@ -0,0 +1,7 @@
+jobmanager.rpc.address: jobmanager
```
Since users have to copy this file anyway, maybe the initial version remains bare-bones and part of the walkthrough is filling it out. Ports and anything that needs to be coordinated with docker to work should be present, but savepoint/checkpoint options for instance.

sjwiesman replied:
At a minimum, since this is an operations walkthrough it should walk through the configs.

knaufk replied:
Good point.

knaufk replied:
I thought about this again (see my post slack). For me this is not so much an operations walkthrough, but the description of a simple, working docker-compose setup to play around with. Therefore I don't have multiple steps for now.

knaufk committed:
[FLINK-12749] [docs] add Flink Cluster Playground documentation

AjaxXu committed:
[FLINK-13161][table-blink-runtime] Fix calculation of numBuckets in BinaryHashBucketArea

The original calcucation may lead to some requested segments are wasted.

This closes #9027

GJL committed:
[hotfix][tests] Declare variables in start_taskmanagers() local

GJL committed:
[hotfix][tests] Improve wait_job_terminal_state() function

Always terminate script if job is in globally terminal state. Assert that
expected terminal state matches actual state.

GJL committed:
[FLINK-13145][runtime, tests] Enable fine grained failover in E2E HA data set test

- Introduce new test job (DataSetFineGrainedRecoveryTestProgram), which waits
  for an  external file to be created before finishing.
- Introduce killing of TMs to HA data set test.
- Reduce JM kills to 2.
- Reduce heartbeat interval and timeout to speed up TM loss detection.

This closes #9060.

twalthr committed:
[hotfix][table-common] Fix minor typos in ObjectIdentifier

twalthr committed:
[hotfix][table-common] Update list of synonyms for logical types

twalthr committed:
[hotfix][table-common] Link timestamp precisions

twalthr committed:
[FLINK-13078][table-common] Simplify serializable string representation for parsers

twalthr committed:
[FLINK-13078][table-common] Add an unresolved user-defined logical type

twalthr committed:
[FLINK-13078][table-common] Add a logical type parser

This adds a parser for all logical types defined in FLIP-37.

This closes #9061.

GJL committed:
[FLINK-13193][tests] Enable custom flink cluster config per test

Enable users to set a cluster configuration in the test spec which will have
precedence over the default configuration.

Remove --ha-storage-dir parameter because the user can set this value in the
test spec cluster configuration now.

This closes #9070.

sjwiesman committed:
[FLINK-13094][state-processor-api] Support iterating over registered timers in the internal timer service

Certain operations require querying the registered timers for a key but
timers are stored TIMESTAMP -> KEY[] which would only support O(n)
reads. Because the timer service sits in the per-record code path we opt
to add forEach*RegisteredTimer methods so consumers can copy data into a
more appropriate data structure to avoid any accidental performance
degredations.

sjwiesman committed:
[FLINK-13094][state-processor-api] Add registered*TimeTimers methods to KeyedStateReaderFunction#Context for querying the registered timers for a given key

This closes #9094.

godfreyhe committed:
[FLINK-13116][table-planner-blink] Fix Catalog statistics is not bridged to blink planner

This closes #9083

Commit:
[hotfix][test] Extract common variables

Commit:
[FLINK-12171][Network] Do not limit the network buffer memory by heap size on the TM side

Commit:
[hotfix][test] Converting fraction to double to improve the precision

lirui-apache committed:
[FLINK-13012][hive] Handle default partition name of Hive table

This closes #9088

Commit:
[hotfix][Network] Rename the terminology

xintongsong committed:
[FLINK-13241][yarn/mesos] Fix Yarn/MesosResourceManager setting managed memory size into wrong configuration instance.

[FLINK-13241][yarn][test] Update YarnResourceManagerTest#testCreateSlotsPerWorker to compute tmCalculatedResourceProfile based on the RM altered configuration.

[FLINK-13241][yarn][test] Update YarnConfigurationITCase to verify that TMs are started with correct managed memory size.

[FLINK-13241][runtime] Calculating and set managed memory size outside of ResourceManager.

[FLINK-13241][rumtime/yarn][test] Move YarnResourceManagerTest#testCreateSlotsPerWorker to ResourceManagerTest#testCreateWorkerSlotProfiles, and update to verify slot profile calculation with determinate managed memory size.

[FLINK-13241][runtime] Move getResourceManagerConfiguration from ResourceManagerFactory to ResourceManagerUtil.

This closes #9246.

Commit:
[FLINK-12765][coordinator] Add the remaining resources to the slot selection strategy

JingsongLi committed:
[FLINK-13257][table-planner-blink] Avoid stream operator implementing BoundedOneInput in blink runner

This closes #9109

zjuwangg committed:
[FLINK-13206]replace 'use database' with 'use' in sql client parser

Replace use database xxx with use xxx in sql client parser.

This closes #9118.

zhuzhurk committed:
[FLINK-13222][docs] Add documentation for failover strategy option

Add detailed failover strategies documentation to doc page

Point links in ZH docs to ZH pages

Use a relative path in docs link to flink pages

This closes #9113.

azagrebin committed:
[FLINK-13407][tests][coordination] Harden StandaloneResourceManagerTest

Get SlotManager.failUnfulfillableRequest in main thread of StandaloneResourceManager
for verification in StandaloneResourceManagerTest

This closes #9126.

wuchong committed:
[FLINK-13216][FLINK-13153][table-planner-blink] Fix Max_Retract and Min_Retract may produce incorrect result

The reason is we didn't set null to acc.max/min which may have an old value when we need to get a new max/min from an empty MapView.
And the old value will be output to downstream instead of a null value. This influences the final result.

This causes many unstable cases, including:
- AggregateITCase.testNestedGroupByAgg
- SplitAggregateITCase.testMinMaxWithRetraction

This closes #9120

1u0 committed:
[FLINK-12858][checkpointing] Fail whole job when in-flight sync savepoint is discarded by a task

knaufk committed:
[FLINK-12749] [examples] added first version of an example for flink cluster playground

docete committed:
[FLINK-13280][table-planner-blink] Revert blink changes in DateTimeUtils, and keep it same as flink version

This closes #9124

zentol committed:
[FLINK-13215][hive] Simplify copying of classpath

zentol committed:
[FLINK-13223][dist] Enable fine-grained recovery by default in distribution

Myasuka committed:
[FLINK-12916][tests] Rework regex

mans2singh committed:
[FLINK-13104][metrics][datadog] Log all failed requests

zentol committed:
[FLINK-11428][tests] Prevent file leak

zjuwangg committed:
[FLINK-13274]Refactor HiveTableSourceTest using HiveRunner

Refactor HiveTableSourceTest to use HiveRunner.

This closes #9130.

godfreyhe committed:
[FLINK-13269][table] Copy RelDecorrelator & FlinkFilterJoinRule to flink planner to fix CALCITE-3169 & CALCITE-3170

This closes #9122

knaufk committed:
[hotfix] upgrade StateMachineExample to universal Kafka connector

zhijiangW committed:
[FLINK-13245][network] Fix the bug of file resource leak while canceling partition request

On producer side the netty handler receives the CancelPartitionRequest for releasing the SubpartitionView resource.
In previous implementation we try to find the corresponding view via available queue in PartitionRequestQueue. But
in reality the view is not always available to stay in this queue, then the view would never be released.

Furthermore the release of ResultPartition/ResultSubpartitions is based on the reference counter in ReleaseOnConsumptionResultPartition,
but while handling the CancelPartitionRequest in PartitionRequestQueue, the ReleaseOnConsumptionResultPartition is never
notified of consumed subpartition. That means the reference counter would never decrease to 0 to trigger partition release,
which would bring file resource leak in the case of BoundedBlockingSubpartition.

In order to fix above two issues, the corresponding view is released via all reader queue instead, and then it would call
ReleaseOnConsumptionResultPartition#onConsumedSubpartition meanwhile to solve this bug.

azagrebin committed:
[hotfix][tests] StandaloneResourceManagerTest extends TestLogger

sjwiesman committed:
[FLINK-13293] [state-processor-api] [build] Add state processor api to opt/ directory in flink-dist

This closes #9133.

bowenli86 committed:
[FLINK-13294][sql client] disable LocalExecutorITCase.testUseCatalogAndUseDatabase

bowenli86 committed:
[FLINK-13296][table] FunctionCatalog.lookupFunction() should check in memory functions if the target function doesn't exist in catalog

Currently the logic to lookup a function is check either the catalog or the in memory function. But the correct logic is to 1st check the catalog, and if the function doesn't exist there, check in memory functions. There should be a resolution order.

This closes #9135.

bowenli86 committed:
[FLINK-13276][hive] write documentation and quickstart for Flink-Hive compatibility

This PR adds documentation and quickstart for Flink-Hive integration.

This closes #13276.

lincoln-lil committed:
[FLINK-13284][table-planner-blink] Correct some builtin functions' return type inference in Blink planner

This closes #9146

1u0 committed:
[hotfix][tests] TimestampITCase: use explicit CheckpointFailureReason to ignore exception

docete committed:
[FLINK-13302][table-planner] Fix CEIL(date) returns the same value with FLOOR(date)

This is a bug that DateTimeUtils.unixDateCeil returns the same value as unixDateFloor does.

This closes #9160

godfreyhe committed:
[FLINK-13266][table] Move OptimizerConfigOptions & ExecutionConfigOptions to table-api-java module

kl0u committed:
[FLINK-13307][tests] Fix SourceStreamTaskTest test instability.

godfreyhe committed:
[FLINK-13168][table] Clarify isBatch/isStreaming/isBounded flag in flink planner and blink planner

kl0u committed:
[FLINK-13275] Fix race condition in finishTask().

godfreyhe committed:
[FLINK-13266] [table] move ExpressionParserException & UnresolvedException to table-common module

godfreyhe committed:
[FLINK-13266] [table] remove definedTimeAttributes file in blink planner

those classes are already in table-common module

JingsongLi committed:
[FLINK-13287][table-planner] Support STREAM_RECORD_TIMESTAMP call in table planner

JingsongLi committed:
[FLINK-13287][table-planner] Support Reinterpret cast call in blink planner

JingsongLi committed:
[FLINK-13287][table-api] Port ExistingField to api-java and use new Expression in FieldComputer

JingsongLi committed:
[FLINK-13287][table-api] Port StreamRecordTimestamp to api-java

This closes #9129.

dianfu committed:
[hotfix] Add directory link-python/apache_flink.egg-info to gitignore

JingsongLi committed:
[FLINK-13315][api-java] Port wmstrategies to api-java-bridge

This closes #9153.

zentol committed:
[FLINK-13316][legal] Update binary licensing

This commit includes the deterministic ordering of NOTICE entries (FLINK-11836), resulting in an abnormally large diff.
It is recommended to view the actual diff via "git diff release-1.8.0..<branch> --patience".

licenses-binary changes and their cause:
flink-ml removal:
- arpack
- machnist
- spire
- jtransforms
- core
flink-shaded-hadoop removal:
- jsch
- paranamer
- xmlenc
- cddlv1.0
- cddlv1.1
flink-shaded update:
- asm
influxdb-reporter:
- influx

The remaining changes are due to the WebUI update.

NOTICE-binary changes:
It is hard to tell which changes this commit specifically does (due to the re-ordering).
Hence, here's a changelog compared to 1.8.0:

add flink-azure-fs-hadoop
update akka
update scala-parser-combinators
update flink-shaded-jackson
add new WebUI
remove flink-shaded-asm (switched fully to flink-shaded-asm-6)
remove flink-ml-uber
remove flink-ml
remove apache camel (previous flink-yarn dependency)
replace flink-python
add flink-shaded-netty-openssl-static
update calcite
update janino

Commit:
[FLINK-13326] Add closeInterruptibly()

Commit:
[FLINK-13326] Use ResourceGuard in NonClosingCheckpointStream

This commit adds the ability for users to indicate to Flink that the
KeyedStateCheckpointOutputStream / OperatorStateCheckpointOutputStream should be closed
at some point later in time. Previously it was expected to be closed within the synchrouns
part of AbstractStreamOperator#snapshotState.
Users might call KeyedStateCheckpointOutputStream#acquireLease(), and at some point
in the future might release the least. Once the number of leases reaches 0, the stream
would be closed.

Commit:
[FLINK-13326] Support async rawState checkpointing

This commit adds the ability for users to obtain a lease for the
rawOperator/rawKeyed output streams, during the synchronous part
of a snapshot, thus preventing these streaming to close.

dianfu committed:
[FLINK-13308][python] Drop the classifier of the flink-python jar

WeiZhong94 committed:
[FLINK-13299][travis][python] fix flink-python failed on Travis because of incompatible virtualenv (#9140)

beyond1920 committed:
[FLINK-13314][table-planner-blink] Correct resultType of some PlannerExpression when operands contains DecimalTypeInfo or BigDecimalTypeInfo in Blink planner

This also fix some minor bugs:
- Fix minor bug in RexNodeConverter when convert between and not between to RexNode.
- Fix minor bug in PlannerExpressionConverter when convert DataType to TypeInformation.

This closes #9152

bowenli86 committed:
[FLINK-13312][hive] move tests for data type mappings between Flink and Hive into its own test class

This PR moves UT for data type mapping between Flink and Hive to its own test class.

This closes #9151.

godfreyhe committed:
[FLINK-13266][table] Port DataView related classes to flink-table-common

godfreyhe committed:
[FLINK-13266][table-runtime-blink] remove Order class from table-runtime-blink

docete committed:
[hotfix][kafka] Fix typo in filter property in pom.xml

This closes #9142

tsreaper committed:
[FLINK-13304][table-runtime-blink] Fix implementation of getString and getBinary method in NestedRow

This closes #9139

Myasuka committed:
[FLINK-13256] Ensure periodical checkpointing continues when regional failover aborts pending checkpoints

This closes #9128.

StefanRRichter committed:
[FLINK-13249][runtime] Fix handling of partition producer responses b… (#9138)

* [FLINK-13249][runtime] Fix handling of partition producer responses by running them with the task's executor

* Review comments

Myasuka committed:
[FLINK-13281][tests] Improve test for aborting pending checkpoints on failover

zentol committed:
[FLINK-13320][catalog][docs] Remove broken link

zentol committed:
[FLINK-13319][catalog][docs] Add chinese version

twalthr committed:
[FLINK-13262][docs] Add documentation for the new Table & SQL API type system

This closes #9161.

tsreaper committed:
[FLINK-13322][table-runtime-blink] Fix serializer snapshot recovery in BaseArray and BaseMap serializers

zentol committed:
[FLINK-12578][build] Use secure MapR repository by default

zentol committed:
[FLINK-12578][build] Add fallback unsafe MapR repository

godfreyhe committed:
[FLINK-13266][table] Port function-related descriptors to flink-table-common

FileSystem/OldCsv/RowtimeValidator/SchemaValidator will be ported in other commits

beyond1920 committed:
[FLINK-13321][table-planner-blink] Fix lateral join udtf with constant parameters doesn't work in blink planner

This closes #9162

tsreaper committed:
[FLINK-13323][table-runtime-blink] Add tests for complex data formats

1u0 committed:
[FLINK-12998][docs] Document optional file systems libs to use plugins loading mechanism

dawidwys committed:
[hotfix][table-api-java] Improve documentation of withBuiltinCatalogName & withBuiltinDatabaseName

dawidwys committed:
[hotfix][table] Unify default catalog & builtin catalog naming

JingsongLi committed:
[FLINK-13286][table-api] Port RowtimeValidator and SchemaValidator to api-java-bridge

This closes #9168

JingsongLi committed:
[FLINK-13286][table-common] Port FileSystem and FileSystemValidator to common

1u0 committed:
[hotfix][docs] Reference file systems documentation from batch/connectors docs

dawidwys committed:
[FLINK-13327][table-planner-blink] Fixed scala 2.12 compilation

bowenli86 committed:
[FLINK-13313][table] create CatalogTableBuilder to support building CatalogTable from descriptors

This PR adds CatalogTableBuilder as a replacement of ExternalCatalogTableBuilder to help users convert table source/sink descriptors to CatalogTable. The gap was mainly discovered when I was writing tests for HiveCatalog to make sure it works as expected to persist Flink generic tables

This closes #9172.

shuai-xu committed:
[FLINK-12038][tests] Harden YARNITCase

Only kill Yarn application if it does not properly terminate.

This closes #9175.

godfreyhe committed:
[FLINK-13266][table-common] Improve comment for FilterableTableTableSource

Flink planner planner will push down PlannerExpressions (which are defined in flink-table-planner module), while Blink planner will push down Expressions. So the implementation for Flink planner and Blink planner should be different and incompatible.

lirui-apache committed:
[FLINK-13210][hive] Hive connector test should dependent on blink planner instead of legacy planner

This closes #9181

Commit:
[FLINK-13117][table-planner-blink] Do not need to backup and restore streamEnv config in BatchExecutor

This closes #9179

zentol committed:
[FLINK-12578][build] Use fallback unsafe MapR repository

hequn8128 committed:
[FLINK-12249][table] Fix type equivalence check problems for Window Aggregates

This closes #9141

WeiZhong94 committed:
[hotfix][docs] Fix loop redirection in /dev/projectsetup/dependencies.html (#9177)

yanghua committed:
[FLINK-13186][clients] Remove unused fields

Commit:
[FLINK-13310][hive][build] Remove shade-plugin configuration

zentol committed:
[FLINK-13255][hive] Skip tests on Java 9

godfreyhe committed:
[FLINK-13266][table] Relocate blink runtime classes to avoid class clashes

1u0 committed:
[FLINK-12998][docs] Extend debugging_classloading documentation with plugins loading case

kaka11chen committed:
[FLINK-13341][table][connectors] StreamTableSink#consumeDataStream returns DataStreamSink when using blink planner.

This closes #9186

rehevkor5 committed:
[FLINK-12595][kinesis] Interrupt thread at right time to avoid deadlock

- Inside testOriginalExceptionIsPreservedWhenInterruptedDuringShutdown,
consumerThread.interrupt() was getting absorbed inside
KinesisDataFetcher's while(running) loop, therefore
TestableKinesisDataFetcherForShardConsumerException's awaitTermination()
wasn't getting interrupted by it. This led to deadlock, with
KinesisDataFetcher waiting on the test code to send the interrupt, and
the test code waiting for KinesisDataFetcher to throw the expected
exception.
- Now, the test code waits until KinesisDataFetcher is inside
awaitTermination() before producing the interrupt, so it can be sure
that the interrupt it produces will be received/handled inside
awaitTermination().

GJL committed:
[FLINK-13345][tests] Dump jstack output for Flink JVMs

Dump the jstack output for all Flink JVMs at the end of each Jepsen test in the
log aggregation phase. This can be helpful for debugging deadlocks.

This closes #9194.

liyafan82 committed:
[FLINK-13221][table-planner-blink] Blink planner should set ScheduleMode to LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST for batch jobs

This closes #9101

godfreyhe committed:
[FLINK-13266][table] Relocate blink planner classes to avoid class clashes

beyond1920 committed:
[FLINK-13318][table-planner-blink] Fix Blink planner tests failing on Scala 2.12 by setting SerialVersionUID to Scala UDFs

This is because:

1. we print the udf's md5 into the node explanation
2. blink planner uses XML to verify plan, the XML content is immutable between different Scala versions
3. the serialization behavior is different between Scala 2.11 and 2.12 when version id is not defined.

That's why we add version id to all existing scala udfs. But this doesn't mean we force users to set this. This is just a workaround for the plan test for different Scala versions. Users can still use a Scala udf without version id in blink planner and flink planner.

This closes #9193

JingsongLi committed:
[FLINK-13289][table-planner-blink] Blink planner should setKeyFields to upsert table sink

This closes #9195

AjaxXu committed:
[FLINK-13351][table-blink-planner] Clean up duplicate case clause for ROW in FlinkTypeFactory.toLogicalType

This closes #9196

tsreaper committed:
[FLINK-13075][table-planner-blink] Project pushdown rule shouldn't require the TableSource return a modified schema in blink planner

This closes #9197

Commit:
[FLINK-13367] Recognize writeReplace in ClosureCleaner

This closes #9201

WeiZhong94 committed:
[FLINK-13368][python] Add Configuration Class for Python Table API to Align with Java.

This closes #9199

ifndef-SleePy committed:
[FLINK-9526][e2e] Fix unstable BucketingSink end-to-end test

Commit:
[FLINK-13369] Track references of already visited object in ClosureCleaner

kl0u committed:
[hotfix] fix checkstyle

JingsongLi committed:
[FLINK-13378][table-planner-blink] Fix SINGLE_VALUE is not correctly supported in blink planner

This closes #9208

JingsongLi committed:
[FLINK-13393][table-planner-blink] Fix generic TableSource doesn't work in blink planner

The reason is that blink-planner use TypeExtractor to extract class from TableSource, and use this class to DataFormatConverter. However, blink should use conversion class of DataType in TableSource#getProducedDataType.

This closes #9211

JingsongLi committed:
[FLINK-13391][table] Fix InputFormatTableSource#getDataStream should not call getReturnType.

danny0405 committed:
[FLINK-13338][table] Make SQL dialect configurable

This closes #9212.

dawidwys committed:
[FLINK-13350][table-api-java] Annotate useCatalog & useDatabase with experimental annotation

dawidwys committed:
[hotfix][table-common] Fix logging in the table source validation procedure

zentol committed:
[FLINK-13394][travis] Use fallback unsafe MapR repository

aljoscha committed:
[FLINK-13397] Remove RollingSink

godfreyhe committed:
[FLINK-13402][table] Copy RelOptCluster to flink planner to fix "RelOptCluster's constructor can not be accessed when connector depends on both planners"

This closes #9215

dawidwys committed:
[FLINK-13279][table-sql-client] Fully qualify sink name in sql-client

This closes #9229.

godfreyhe committed:
[FLINK-13403][table-planner-blink] Fix some classes' package name is not updated after class relocation

This closes #9216

zentol committed:
[FLINK-13398][hive][build] Exclude jdk.tools

Specify a direct exclusion for jdk.tools on the root dependencies, since dependencyManagement entries for transitive are ignored by the shade-plugin artifact resolution.

azagrebin committed:
[FLINK-13169][tests][coordination] Test TaskExecutor failures in fine-grained recovery IT case

The BatchFineGrainedRecoveryITCase can be extended with an additional test failure strategy which abruptly shuts down the task executor. This leads to the loss of all previously completed and the in-progress mapper result partitions. The fail-over strategy should restart the current in-progress mapper which will get the PartitionNotFoundException because the previous result becomes unavailable and the previous mapper has to be restarted as well. The same should happen subsequently with all previous mappers. When the source is recomputed, all mappers has to be restarted again to recalculate their lost results.

zhuzhurk committed:
[FLINK-13055][runtime] Leverage PartitionTracker for checking partition availability

zentol committed:
[hotfix][tests] Delete tmp directory after cluster has shut down

AjaxXu committed:
[FLINK-13353][table-planner] Remove 2 args constructor in REPLACE expression

This closes #9218

ifndef-SleePy committed:
[FLINK-13376][datastream] ContinuousFileReaderOperator should respect semantics of BoundedOneInput

zentol committed:
[FLINK-12901][docs] Update Hadoop build instructions

WeiZhong94 committed:
[FLINK-13409][python] Supported java UDFs in python API.

This closes #9222

becketqin committed:
[hotfix] Fix static inner class name collision in KafkaConsumerTestBase.

godfreyhe committed:
[FLINK-13347][table-planner] should handle SEMI/ANTI JoinRelType in switch case

This closes #9227.

wuchong committed:
[FLINK-13116][table-planner-blink] Exclude RAT for explain and digest resource files

lirui-apache committed:
[FLINK-13423][hive] Unable to find function in hive 1

This closes #9279

kl0u committed:
[FLINK-13440] Report reason when failing job due to checkpoint failure.

lirui-apache committed:
[FLINK-13424][hive] HiveCatalog should add hive version in conf

To avoid overriding the hive version users specify in the yaml file.

This closes #9232.

carp84 committed:
[FLINK-13228][tests][filesystems] Harden HadoopRecoverableWriterTest

Currently test cases will fail when trying to close the output stream if all data written
but ClosedByInterruptException occurs at the ending phase. This commit fixes it.

This closes #9235

zhijiangW committed:
[FLINK-13245][network] Make subpartition consumption notification independant

xintongsong committed:
[FLINK-13408][runtime] Let StandaloneResourceManager start startup period upon granting leadership

During the startup period, the StandaloneResourceManager won't fail slot requests if it cannot allocate new
containers.

[FLINK-13408][runtime][test] Add StandaloneResourceManagerTest#testStartUpPeriodAfterLeadershipSwitch validates that StandaloneResourceManager applies a startup period whenever the leadership is acquired.

[FLINK-13408][runtime] StandaloneResourceManager#startStartupPeriod uses RpcEndpoint#scheduleRunAsync.

This closes #9242.

zhijiangW committed:
[FLINK-13245][network] Remove redundant bookkeeping for already canceled input channel IDs

azagrebin committed:
[FLINK-13435] Remove ShuffleDescriptor.ReleaseType and make release semantics fixed per partition type

In a long term we do not need auto-release semantics for blocking (persistent) partition. We expect them always to be released externally by JM and assume they can be consumed multiple times.

The pipelined partitions have always only one consumer and one consumption attempt. Afterwards they can be always released automatically.

ShuffleDescriptor.ReleaseType was introduced to make release semantics more flexible but it is not needed in a long term.

FORCE_PARTITION_RELEASE_ON_CONSUMPTION was introduced as a safety net to be able to fallback to 1.8 behaviour without the partition tracker and JM taking care about blocking partition release. We can make this option specific for NettyShuffleEnvironment which was the only existing shuffle service before. If it is activated then the blocking partition is also auto-released on a consumption attempt as it was before. The fine-grained recovery will just not find the partition after the job restart in this case and will restart the producer.

zentol committed:
[hotfix][coordination] Check whether partition set to track is empty

zentol committed:
[FLINK-13374][scala][build] Set -Xss2m when compiling scala

hongtao12310 committed:
[FLINK-13431][hive] Fix nameNode HA configuration was not loaded when running HiveConnector on Yarn

This closes #9237

tisonkun committed:
[FLINK-12179][coordination] Remove legacy class Instance

yanghua committed:
[FLINK-13158][rest] Remove WebMonitor interface

zentol committed:
[FLINK-13387][WebUI] Fix log download for old UI

kl0u committed:
[FLINK-12858] Change the CHECKPOINT_COORDINATOR_SHUTDOWN to be not pre-flight.

zentol committed:
[FLINK-12590][docs] Use HTTPS for all Flink links

kl0u committed:
[FLINK-13440] Move checkpoint failure logic from scheduler to failure manager

kl0u committed:
[FLINK-13440] Add test for FLINK-12858

twalthr committed:
[FLINK-13429][table-common] Fix BoundedOutOfOrderTimestamps watermark strategy

This closes #9241.

danny0405 committed:
[FLINK-13335][sql-parser] Bring the SQL CREATE TABLE DDL closer to FLIP-37

This brings the SQL DDL closer to FLIP-37. However, there are a couple of
known limitations.

Currently unsupported features:
- INTERVAL
- ROW with comments
- ANY
- NULL
- NOT NULL/NULL for top-level types
- ignoring collation/charset
- VARCHAR without length (=VARCHAR(1))
- TIMESTAMP WITH TIME ZONE
- user-defined types
- data types in non-DDL parts (e.g. CAST(f AS STRING))

This closes #9243.

kl0u committed:
[FLINK-13396] Deprecate the BucketingSink

StephanEwen committed:
[hotfix][tests] Pull TestTaskBuilder out of TaskTest

StephanEwen committed:
[hotfix][tests] Make NoOpPartitionProducerStateChecker a public reusable test implementation.

StephanEwen committed:
[hotfix][tests] Reduce some mocking in TestTaskBuilder

StephanEwen committed:
[FLINK-13325][test] Add test for FLINK-13249

FLINK-13249 was a bug where a deadlock occurred when the network thread got blocked on a lock
while requesting partitions to be read by remote channels. The test mimicks that situation
to guard the fix applied in an earlier commit.

wuchong committed:
[FLINK-13375][table] Improve config names in ExecutionConfigOptions and OptimizerConfigOptions

This closes #9203

wuchong committed:
[FLINK-13447][table] Change default planner to legacy planner instead of any one

This closes #9249

StephanEwen committed:
[FLINK-13451][tests] Remove use of Unsafe.defineClass() from CommonTestUtils

The method Unsafe.defineClass() is removed in Java 11. To support Java 11, we rework the method
"CommonTestUtils.createClassNotInClassPath()" to use a different mechanism.

This commit now writes the class byte code out to a temporary file and create a new URLClassLoader that
loads the class from that file.  That solution is not a complete drop-in replacement, because it cannot
add the class to an existing class loader, but can only create a new pair of (classloader & new-class-in-that-classloader).
Because of that, the commit also adjusts the existing tests to work with that new mechanism.

This closes #9251

lirui-apache committed:
[FLINK-13352][docs] Using hive connector with hive-1.2.1 needs libfb303 jar

This closes #9223

lirui-apache committed:
[FLINK-13427][hive] HiveCatalog's createFunction fails when function name has upper-case characters

This closes #9254

zentol committed:
[FLINK-13453][build] Bump shade plugin to 3.2.1

zentol committed:
[FLINK-13454][build] Bump japicmp jaxb dependencies

zentol committed:
[FLINK-13455][build] Move jdk.tools exclusions out of dependency management

zentol committed:
[FLINK-13456][build] Bump lombok to 1.16.22

twalthr committed:
[FLINK-13458][table] ThreadLocalCache clashes for Blink planner

This closes #9257.

twalthr committed:
[table-api-java] Postpone check for Blink planner in StreamTableEnvironment

twalthr committed:
[FLINK-13273][table-planner-blink] Allow retrieving StreamGraph from Blink executor

twalthr committed:
[hotfix][table-planner-blink] Update Blink's auto-complete hints

twalthr committed:
[hotfix][table-planner-blink] Update Blink's ExpressionReducer classloading to Flink's behavior

zentol committed:
[FLINK-13464][build] Bump powermock to 2.0.2

zentol committed:
[FLINK-13465][build] Bump javassist to 3.24.0-GA

yngwiewang committed:
[hotfix] Fix typo in api_concepts.md

dawidwys committed:
[FLINK-13399][table] Create two separate table uber jars for old and blink planners

dawidwys committed:
[FLINK-13399][table][docs] Updated the required dependencies section of Table API documentation

twalthr committed:
[FLINK-13273][sql-client] Allow switching planners in SQL Client

This closes #9266.

kl0u committed:
[hotfix] Rename flag in CheckpointFailureReason to preFlight

azagrebin committed:
[hotfix][tests] Remove setting the default value of force-release-on-consumption

azagrebin committed:
[hotfix][network] fix codestyle issues in ResultPartitionFactory

azagrebin committed:
[hotfix][network] Annotate NettyShuffleDescriptor#PartitionConnectionInfo with @FunctionalInterface

azagrebin committed:
[hotfix][network] fix codestyle issues in NettyShuffleMaster

azagrebin committed:
[hotfix] fix codestyle issues in ShuffleDescriptor

azagrebin committed:
[hotfix][tests] Make PartitionTestUtils enum singleton and fix codestyle

azagrebin committed:
[hotfix][tests] fix codestyle issues in ResultPartitionBuilder

azagrebin committed:
[hotfix][tests] fix codestyle issues in ResultPartitionFactoryTest

azagrebin committed:
[hotfix][tests] fix codestyle issues in NettyShuffleEnvironmentBuilder

azagrebin committed:
[hotfix][tests] fix codestyle issues in NettyShuffleDescriptorBuilder

zentol committed:
[FLINK-13511][kafka][build] Add test jaxb dependency

ifndef-SleePy committed:
[FLINK-9900][tests] Fix unstable ZooKeeperHighAvailabilityITCase

This closes #9900.

HuangXingBo committed:
[hotfix][python] Change "flink-python-" to "flink-python" for the change  of artifact of flink-python module (#9270)

StephanEwen committed:
[hotfix][tests] Refactor MapR FS Tests

twalthr committed:
[hotfix][sql-client] Add USE CATALOG/USE to CLI help

xintongsong committed:
[hotfix][runtime] Extract SlotManager interface

Rename SlotManager -> SlotManagerImpl and let it implement the SlotManager interface

ifndef-SleePy committed:
[FLINK-13376][datastream] Unify all no operation sinks with DiscardingSink

becketqin committed:
[FLINK-13226][connector/kafka] Fix deadlock between producer closure and transaction commit in Kafka connector 0.11

This patch fixes a race condition between the checkpointing thread and main thread.
The sequence causing the deadlock is the following:

    1. In FlinkKafkaProducer, the main thread encounters a problem and closes all the producer
       to start failover.
    2. The previous checkpoint has completed, so the checkpointing thread grabs the checkpoint
       lock and tries to commit the transaction on the producer that has been closed in step 1.
       This commit will never succeed due to KAFKA-6635. So the checkpoint thread blocks forever.
    3. In StreamTask, the main thread will eventually try to release all the record writer.
       To do that, it attempts to grab the checkpoint lock which is hold by checkpoint thread in
       step 2 and will never be released. So the main thread also blocks forever.

KAFKA-6635 has been fixed in Kafka 2.3.0. But Flink 1.9 does not rely on that yet, and we also
support Kafka 0.11. So we are just going to fix on the Flink side first. The solution is to make
sure that in FlinkKafkaProducer any operation relying on the underlying sender thread to finish
throws an exception if the producer is closed.

becketqin committed:
[FLINK-13226][connector/kafka] Fix deadlock between producer closure and transaction commit in the universal Kafka connector.

This patch fixes a race condition between the checkpointing thread and main thread.
The sequence causing the deadlock is the following:

    1. In FlinkKafkaProducer, the main thread encounters a problem and closes all the producer
       to start failover.
    2. The previous checkpoint has completed, so the checkpointing thread grabs the checkpoint
       lock and tries to commit the transaction on the producer that has been closed in step 1.
       This commit will never succeed due to KAFKA-6635. So the checkpoint thread blocks forever.
    3. In StreamTask, the main thread will eventually try to release all the record writer.
       To do that, it attempts to grab the checkpoint lock which is hold by checkpoint thread in
       step 2 and will never be released. So the main thread also blocks forever.

KAFKA-6635 has been fixed in Kafka 2.3.0. But Flink 1.9 does not rely on that yet, So we are just
going to fix on the Flink side first. The solution is to make sure that in FlinkKafkaProducer any
operation relying on the underlying sender thread to finish throws an exception if the producer
is closed.

StephanEwen committed:
[FLINK-13499][maprfs] Handle MapR dependency purely through reflection

This allows us to remove the MapR dependency from the module.
The MapR maven dependency has frequently caused issues.

wuchong committed:
[FLINK-13290][table-planner-blink] SinkCodeGenerator should not compare row type field names

wuchong committed:
[FLINK-13290][hbase] Enable blink planner for integration tests of HBase

This commit combines HBaseTableSourceITCase and HBaseLookupFunctionITCase and HBaseConnectorITCase into one class.
This can save much cluster initialization time for us.

This closes #9275

tillrohrmann committed:
[hotfix] Make TestingSlotManager#setFailUnfulfillableRequestConsumer non-nullable

tillrohrmann committed:
[hotfix] Remove polling loops from StandaloneResourceManagerTest

tillrohrmann committed:
[hotfix] Rename TestingSlotManagerFactory into TestingSlotManagerBuilder

tillrohrmann committed:
[hotfix] Fix checkstyle violations in StandaloneResourceManagerTest

sunjincheng121 committed:
[FLINK-13488][Python] Remove python 3.3/3.4 support (#9278)

twalthr committed:
[FLINK-13399][legal] Update NOTICE-binary for new table uber jars

This closes #9261.

twalthr committed:
[hotfix][table] Add missing annotations to factories

Commit:
[FLINK-13494][table-planner-blink] Remove source and sink parallelism configurations

This closes #9277

zhuzhurk committed:
[FLINK-13421] Exclude releasing root slots from slot allocation

Make MultiTaskSlot not available for allocation when it’s releasing children
to avoid ConcurrentModificationException.

This closes #9288.

dianfu committed:
[hotfix] [travis] Fix the python travis failure (#9286)

zjuwangg committed:
[FLINK-12966][hive] finalize package name of Hive table source/sink

finalize package name of Hive table source/sink from org.apache.flink.batch.connector to org.apache.flink.connector.

This closes #9290.

azagrebin committed:
[hotfix][tests] fix codestyle issues in NettyShuffleEnvironmentConfiguration

wuchong committed:
[FLINK-13290][table-api] Add method to check LogicalType compatible

Add a areTypesCompatible() method to LogicalTypeChecks. This will compare two LogicalTypes without field names and other logical attributes (e.g. description, isFinal).

gaoyunhaii committed:
[FLINK-13487][tests] Wait until TE has registerd at JM

zentol committed:
[FLINK-10725][build] Initialize JDK 11 profile

kl0u committed:
[FLINK-13434][e2e] Change the test_resume_savepoint to use stop-with-savepoint.

kl0u committed:
[FLINK-13434][e2e] Change test_stateful_stream_job_upgrade to use stop-with-savepoint.

azagrebin committed:
[hotfix][network] Simplify ResultPartitionFactory.createSubpartitions based on ResultPartitionType.isBlocking

kl0u committed:
[hotfix][e2e] Harden test_resume_savepoint e2e test.

kl0u committed:
[hotfix][e2e] Harden test_stateful_stream_job_upgrade e2e test.

tsreaper committed:
[table-api-java] Change the default value of table.exec.sort.default-limit to -1

Fokko committed:
[FLINK-11321] Clarify NPE on fetching nonexistent topic (#7487)

[FLINK-11321][connector/kafka] Throw explicit exception on fetching nonexistent topic.

Before the fix, NPE will be thrown if a consumer is trying to fetch from a nonexistent topic. The fix throws a RuntimeException with a clear message.

tsreaper committed:
[FLINK-13436][e2e] Add TPC-H queries as E2E tests

This closes #9312

dawidwys committed:
[FLINK-13504][table] Fixed shading issues in table modules.

Properly include the flink-sql-parser into flink-planner* modules shaded
jars. Removed the flink-sql-parser module from flink-table-uber* jars as
it is already included in the flink-planner* modules.

This closes #9313.

dawidwys committed:
[hotfix][table-sql-parser] Removed unnecessary dependency on guava.

WeiZhong94 committed:
[FLINK-12704][python] Enable the configuration of using blink planner.

This closes #9314

ifndef-SleePy committed:
[FLINK-13491][datastream] correctly support endInput in AsyncWaitOperator

Before completing endInput operator has to make sure that all of the records were emitted.

zentol committed:
[FLINK-13371][coordination] Prevent leaks of blocking partitions

wuchong committed:
[FLINK-13543][table-planner-blink] Enable reuse forks for integration tests in blink planner

This closes #9326

tsreaper committed:
[FLINK-13546][travis] Add TPC-H E2E test in travis cron job

This closes #9330

wuchong committed:
[FLINK-13544][connectors] Set parallelism of table sink operator to input transformation parallelism

This closes #9332

dawidwys committed:
[hotfix][table-planner] Added joda.convert to planner jar
